File: ModelIO_Bindings/ModelIO_Bindings/ModelIOMeshes.swift
//
//  ModelIOMeshes.swift
//  ModelIO_Bindings
//
//  Created by Taanish Reja on 1/8/25.
//

import ModelIO
import MetalKit
import Metal

// MDL Mesh

// create
// sphere with extent
// (sphereWithExtent: <#T##vector_float3#>, segments: <#T##vector_uint2#>, inwardNormals: <#T##Bool#>, geometryType: <#T##MDLGeometryType#>, allocator: <#T##MDLMeshBufferAllocator?#>)

func pointerToSIMD<T: SIMD>(
    pointer: UnsafePointer<T.Scalar>
) -> T where T.Scalar: Numeric {
    let simdVector = T((0..<T.scalarCount).map { pointer[$0] });
    return simdVector;
}

public func createSphereMDLMesh(
    sphereWithExtentPtr: UnsafePointer<Float>,
    segmentsPtr: UnsafePointer<UInt32>,
    inwardNormals: Bool,
    rawGeometryType: Int,
    allocatorPtr: UnsafeMutableRawPointer
    
) -> UnsafeMutableRawPointer
{
    let allocator = Unmanaged<MTKMeshBufferAllocator>.fromOpaque(allocatorPtr).takeUnretainedValue();
    let geometryType = MDLGeometryType(rawValue: rawGeometryType)!;
    let mdlMesh = MDLMesh(sphereWithExtent: pointerToSIMD(pointer: sphereWithExtentPtr),
                          segments: pointerToSIMD(pointer: segmentsPtr),
                          inwardNormals: inwardNormals,
                          geometryType: geometryType,
                          allocator: allocator);
    let mdlMeshPtr = UnsafeMutableRawPointer(Unmanaged.passRetained(mdlMesh).toOpaque());
    return mdlMeshPtr;
}


// box with extent
//(boxWithExtent: <#T##vector_float3#>, segments: <#T##vector_uint3#>, inwardNormals: <#T##Bool#>, geometryType: <#T##MDLGeometryType#>, allocator: MDLMeshBufferAllocator?)

public func createBoxMDLMesh(
    boxWithExtentPtr: UnsafePointer<Float>,
    segmentsPtr: UnsafePointer<UInt32>,
    inwardNormals: Bool,
    rawGeometryType: Int,
    allocatorPtr: UnsafeMutableRawPointer
) -> UnsafeMutableRawPointer
{
    let allocator = Unmanaged<MTKMeshBufferAllocator>.fromOpaque(allocatorPtr).takeUnretainedValue();
    let geometryType = MDLGeometryType(rawValue: rawGeometryType)!;
    let mdlMesh = MDLMesh(boxWithExtent: pointerToSIMD(pointer: boxWithExtentPtr),
                          segments: pointerToSIMD(pointer: segmentsPtr),
                          inwardNormals: inwardNormals,
                          geometryType: geometryType,
                          allocator: allocator);
    let mdlMeshPtr = UnsafeMutableRawPointer(Unmanaged.passRetained(mdlMesh).toOpaque());
    return mdlMeshPtr;
}

public func setMDLMeshVertexDescriptor(mdlMeshPtr: UnsafeMutableRawPointer,
                                       mdlVertexDescriptorPtr: UnsafeMutableRawPointer)
{
    let mdlMesh = Unmanaged<MDLMesh>.fromOpaque(mdlMeshPtr).takeUnretainedValue();
    let mdlVertexDescriptor = Unmanaged<MDLVertexDescriptor>.fromOpaque(mdlVertexDescriptorPtr).takeUnretainedValue();
    mdlMesh.vertexDescriptor = mdlVertexDescriptor;
}

// release
public func releaseMDLMesh(mdlMeshPtr: UnsafeMutableRawPointer)
{
    let mdlMesh = Unmanaged<MDLMesh>.fromOpaque(mdlMeshPtr);
    mdlMesh.release();
}

// MTKMesh
// getVertexBuffers: [MTKMeshBuffer]
// getVertexDescriptor: MDLVertexDescriptor
// var submeshes: [MTKSubmesh]
// var vertexCount: Int
// c++ equivalents created on construction by calling private helpers

// create
// (MDLMesh, device)
public func createMTKMesh(mdlMeshPtr: UnsafeMutableRawPointer, devicePtr: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer?
{
    let device = Unmanaged<MTLDevice>.fromOpaque(mdlMeshPtr).takeUnretainedValue();
    let mdlMesh = Unmanaged<MDLMesh>.fromOpaque(mdlMeshPtr).takeUnretainedValue();
    
    guard let mtkMesh = try? MTKMesh(mesh: mdlMesh, device: device)
    else {
        return nil;
    };
    
    let mtkMeshPtr = UnsafeMutableRawPointer(Unmanaged.passRetained(mtkMesh).toOpaque());
    
    return mtkMeshPtr;
}

public func getMTKMeshVertexCount(mdlMeshPtr: UnsafeMutableRawPointer) -> Int
{
    let mdlMesh = Unmanaged<MDLMesh>.fromOpaque(mdlMeshPtr).takeUnretainedValue();
    return mdlMesh.vertexCount;
}

// MTKMeshBuffer
// MTLVertexBuffer*
// int offset
// get Vertex buffers
public func countVertexBuffers(
    mtkMeshPtr: UnsafeMutableRawPointer
) -> Int
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr).takeUnretainedValue();
    let vertexBuffers = mtkMesh.vertexBuffers;
    return vertexBuffers.count;
}

public func extractMeshBuffers
(
    mtkMeshPtr: UnsafeMutableRawPointer,
    mtkVertexBuffersPtr: UnsafeMutablePointer<UnsafeMutableRawPointer>,
    offsetsPtr: UnsafeMutablePointer<Int>
)
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr).takeUnretainedValue();
    let vertexBuffers = mtkMesh.vertexBuffers;
    for (index, vertexBuffer) in vertexBuffers.enumerated() {
        mtkVertexBuffersPtr[index] = UnsafeMutableRawPointer(Unmanaged.passUnretained(vertexBuffer.buffer).toOpaque());
        offsetsPtr[index] = vertexBuffer.offset;
    }
}

// MTKSubmesh
//var primitiveType: MTLPrimitiveType
//var indexType: MTLIndexType
//var indexBuffer: MTKMeshBuffer
//var indexCount: Int
public func countSubmeshes(
    mtkMeshPtr: UnsafeMutableRawPointer
) -> Int
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr).takeUnretainedValue();
    let submeshes = mtkMesh.submeshes;
    return submeshes.count;
}


public func extractSubmeshPrimitiveTypes(
    mtkMeshPtr: UnsafeMutableRawPointer,
    primitiveTypesPtr: UnsafeMutablePointer<UInt>
)
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr).takeUnretainedValue();
    let submeshBuffers = mtkMesh.submeshes;
    for (index, submesh) in submeshBuffers.enumerated() {
        primitiveTypesPtr[index] = submesh.primitiveType.rawValue;
    }
}

public func extractSubmeshIndexTypes(
    mtkMeshPtr: UnsafeMutableRawPointer,
    indexTypesPtr: UnsafeMutablePointer<UInt>
)
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr).takeUnretainedValue();
    let submeshBuffers = mtkMesh.submeshes;
    for (index, vertexBuffer) in submeshBuffers.enumerated() {
        indexTypesPtr[index] = vertexBuffer.indexType.rawValue;
    }
}

public func extractSubmeshIndexBuffers
(
    mtkMeshPtr: UnsafeMutableRawPointer,
    submeshBufffersPtr: UnsafeMutablePointer<UnsafeMutableRawPointer>,
    offsetsPtr: UnsafeMutablePointer<Int>
)
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr).takeUnretainedValue();
    let submeshBuffers = mtkMesh.submeshes;
    for (index, submesh) in submeshBuffers.enumerated() {
        submeshBufffersPtr[index] = UnsafeMutableRawPointer(Unmanaged.passUnretained(submesh.indexBuffer.buffer).toOpaque());
        offsetsPtr[index] = submesh.indexBuffer.offset;
    }
}

public func extractSubmeshIndexCounts
(
    mtkMeshPtr: UnsafeMutableRawPointer,
    submeshIndexCounts: UnsafeMutablePointer<Int>
)
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr).takeUnretainedValue();
    let submeshes = mtkMesh.submeshes;
    for (index, submesh) in submeshes.enumerated() {
        submeshIndexCounts[index] = submesh.indexCount;
    }
}

// release
public func releaseMTKMesh(mtkMeshPtr: UnsafeMutableRawPointer)
{
    let mtkMesh = Unmanaged<MTKMesh>.fromOpaque(mtkMeshPtr);
    mtkMesh.release();
}

File: ModelIO_Bindings/ModelIO_Bindings/ModelIOMeshes.cpp
//
//  ModelIOMeshes.cpp
//  ModelIO_Bindings
//
//  Created by Taanish Reja on 1/9/25.
//

#include "ModelIOMeshes.h"

namespace ModelIO {
    MDLMesh::MDLMesh(Sphere sphere, MTKMeshBufferAllocator& allocator):
        m_mesh{std::make_shared<SwiftObject>(
                ModelIO_Bindings::createSphereMDLMesh(
                sphere.sphereWithExtent.data(),
                sphere.segments.data(),
                sphere.inwardsNormal,
                static_cast<uint32_t>(sphere.geometryType),
                allocator.getAllocator()->get()
            ),
            ModelIO_Bindings::releaseMDLMesh
        )}
    {}

    MDLMesh::MDLMesh(Box box, MTKMeshBufferAllocator& allocator):
        m_mesh{std::make_shared<SwiftObject>(
                ModelIO_Bindings::createSphereMDLMesh(
                box.boxWithExtent.data(),
                box.segments.data(),
                box.inwardsNormal,
                static_cast<uint32_t>(box.geometryType),
                allocator.getAllocator()->get()
            ),
            ModelIO_Bindings::releaseMDLMesh
        )}
    {
        
    }
    
    MDLMesh::~MDLMesh() = default;

    SwiftObject* MDLMesh::retrieveMesh()
    {
        return m_mesh.get();
    }
    
    void MDLMesh::setVertexDescriptor(MDLVertexDescriptor& vertexDescriptor)
    {
        ModelIO_Bindings::setMDLMeshVertexDescriptor(m_mesh->get(), vertexDescriptor.getVertexDescriptor()->get());
        m_vertexDescriptor = vertexDescriptor;
    }

    MDLVertexDescriptor& MDLMesh::getVeretexDescriptor()
    {
        return m_vertexDescriptor;
    }

    MTKMeshBuffer::MTKMeshBuffer(MTL::Buffer* buffer, long offset):
        m_buffer{NS::RetainPtr(buffer)}, m_offset{offset} // retain does not take ownership, transfer does
    {}

    MTL::Buffer* MTKMeshBuffer::getBuffer()
    {
        return m_buffer.get();
    }

    long MTKMeshBuffer::getOffset() {
        return m_offset;
    }

    MTKMeshBuffer::~MTKMeshBuffer() = default;

    MTKSubmesh::MTKSubmesh(MTL::PrimitiveType primitiveType,
                      MTL::IndexType indexType,
                      const MTKMeshBuffer& indexBuffer,
                      long indexCount):
        m_primitiveType(primitiveType),
        m_indexType{indexType},
        m_indexBuffer{indexBuffer},
        m_indexCount{indexCount}
    {}

    MTKSubmesh::~MTKSubmesh() = default;


    MTL::PrimitiveType MTKSubmesh::getPrimitiveType()
    {
        return m_primitiveType;
    }

    MTL::IndexType MTKSubmesh::getIndexType()
    {
        return m_indexType;
    }

    MTKMeshBuffer& MTKSubmesh::getIndexBuffer()
    {
        return m_indexBuffer;
    }

    long MTKSubmesh::getIndexCount()
    {
        return m_indexCount;
    }

    MTKMesh::MTKMesh(MDLMesh& mesh, MTL::Device* device):
    m_mesh{std::make_shared<SwiftObject>(
        ModelIO_Bindings::createMTKMesh(mesh.retrieveMesh()->get(), static_cast<void*>(device)),
        ModelIO_Bindings::releaseMTKMesh
    )}
    {
        m_vertexCount = ModelIO_Bindings::getMTKMeshVertexCount(m_mesh->get());
        constructVertexBuffers();
        constructSubmeshes();
    }

    void MTKMesh::constructVertexBuffers()
    {
        long vertexBufferCount = ModelIO_Bindings::countVertexBuffers(m_mesh->get());
        void* vertexBuffers[vertexBufferCount];
        long offsets[vertexBufferCount];
        ModelIO_Bindings::extractMeshBuffers(m_mesh->get(), vertexBuffers, offsets);
        for (int i = 0; i < vertexBufferCount; ++i) {
            m_vertexBuffers.push_back({static_cast<MTL::Buffer*>(vertexBuffers[i]), offsets[i]});
        }
    }

    // TODO
    void MTKMesh::constructSubmeshes()
    {
        long submeshCount = ModelIO_Bindings::countSubmeshes(m_mesh->get());
        
        unsigned long primitiveTypes[submeshCount];
        unsigned long indexTypes[submeshCount];
        void* submeshBuffers[submeshCount];
        long submeshBufferOffsets[submeshCount];
        long indexCounts[submeshCount];
        
        ModelIO_Bindings::extractSubmeshPrimitiveTypes(m_mesh->get(), primitiveTypes);
        ModelIO_Bindings::extractSubmeshIndexTypes(m_mesh->get(), indexTypes);
        ModelIO_Bindings::extractSubmeshIndexBuffers(m_mesh->get(), submeshBuffers, submeshBufferOffsets);
        ModelIO_Bindings::extractSubmeshIndexCounts(m_mesh->get(), indexCounts);
        
        for (int i = 0; i < submeshCount; ++i) {
            m_submeshes.push_back(MTKSubmesh{
                                    static_cast<MTL::PrimitiveType>(primitiveTypes[i]),
                                    static_cast<MTL::IndexType>(indexTypes[i]),
                                    {static_cast<MTL::Buffer*>(submeshBuffers[i]), submeshBufferOffsets[i]},
                                    indexCounts[i]});
        }
        
    }
        
    MTKMesh::~MTKMesh() = default;
        
    std::vector<MTKMeshBuffer>& MTKMesh::getVertexBuffers()
    {
        return m_vertexBuffers;
    }
        
    std::vector<MTKSubmesh>& MTKMesh::getSubmeshes()
    {
        return m_submeshes;
    }
    
    SwiftObject* MTKMesh::retrieveMesh()
    {
        return m_mesh.get();
    }
}
